libraries/chain/include/eosio/chain/combined_database.hpp

```diff
@@ -14,77 +36,88 @@
 
 namespace eosio { namespace chain {
 
-   struct combined_session {
+   using controller_index_set =
+         index_set<account_index, account_metadata_index, account_ram_correction_index, global_property_multi_index,
+                   protocol_state_multi_index, dynamic_global_property_multi_index, block_summary_multi_index,
+                   transaction_multi_index, generated_transaction_multi_index, table_id_multi_index, code_index,
+                   database_header_multi_index, kv_db_config_index, kv_index>;
+
+   using contract_database_index_set = index_set<key_value_index, index64_index, index128_index, index256_index,
+                                                 index_double_index, index_long_double_index>;
+
+   class combined_session {
+    public:
+      combined_session() = default;
+
+      combined_session(chainbase::database& cb_database);
+
+      combined_session(chainbase::database& cb_database, b1::chain_kv::undo_stack& kv_undo_stack);
+
+      combined_session(combined_session&& src) noexcept;
+
+      ~combined_session() { undo(); }
+
+      combined_session& operator=(const combined_session& src) = delete;
+
+      void push();
+
+      void squash();
+
+      void undo();
+
+    private:
       std::unique_ptr<chainbase::database::session> cb_session    = {};
       b1::chain_kv::undo_stack*                     kv_undo_stack = {};
+   };
 
-      combined_session(chainbase::database& cb_database, b1::chain_kv::undo_stack& kv_undo_stack)
-          : kv_undo_stack{ &kv_undo_stack } {
-         try {
-            try {
-               cb_session = std::make_unique<chainbase::database::session>(cb_database.start_undo_session(true));
-               kv_undo_stack.push(false);
-            }
-            FC_LOG_AND_RETHROW()
-         }
-         CATCH_AND_EXIT_DB_FAILURE()
-      }
+   class combined_database {
+    public:
+      combined_database(backing_store_type backing_store, chainbase::database& chain_db,
```

Comment:
Some of the same readability concerns as the others.

Reply:
We need to come together and unify and improve our `.clang-format` files.

Reply:
As a code base and group, we've treated any of the clang-format stuff as purely optional and only to be applied to the code changes that are being made.  Also, if those rules result in bad formatting, then those that do use clang-format will just fix it up as they go along.  There are two issues with clang-format, 1) our codebase is not consistent and we don't particularly care about that. 2) it can format things terribly and there are certain rules that you are just not going to be able to do (and I really don't want to see disable clang formats all over the place).  

For formatting the code, there are only a few rules you need.  Stay consistent with what is around it and make it as readable as possible, at the end of the day that is all we care about.

Reply:
I changed the formatting manually.